src/ServerConnectionAcceptor.cpp

```diff
@@ -159,26 +156,55 @@ void ServerConnectionAcceptor::processFrame(
     default: {
       transport->outputFrameOrEnqueue(
           frameSerializer->serializeOut(Frame_ERROR::unexpectedFrame()));
-      closeAndRemoveConnection(
-          connectionHandler,
-          std::move(transport),
-          std::runtime_error("invalid"));
+      error("Invalid frame received instead of SETUP/RESUME");
       break;
     }
   }
 }
 
+////////////////////////////////////////////////////////////////////////////////
+
 void ServerConnectionAcceptor::accept(
     std::unique_ptr<DuplexConnection> connection,
     std::shared_ptr<ConnectionHandler> connectionHandler) {
   auto transport = std::make_shared<FrameTransport>(std::move(connection));
   auto processor = std::make_shared<OneFrameProcessor>(
       *this, transport, std::move(connectionHandler));
-  connections_.insert(transport);
-  // transport can receive frames right away
+
+  {
+    auto locked = connections_.lock();
+
+    if (shutdown_) {
+      return;
+    }
+
+    locked->insert(transport);
+  }
+
+  // Transport can receive frames right away.
   transport->setFrameProcessor(std::move(processor));
 }
 
+void ServerConnectionAcceptor::stop() {
+  {
+    auto locked = connections_.lock();
+
+    shutdown_.emplace();
+
+    if (locked->empty()) {
+      return;
+    }
+
+    for (auto& connection : *locked) {
+      connection->close(std::runtime_error("ServerConnectionAcceptor closed"));
```

Comment:
Is that because of not having any guarantees about the subscriber/subscription pair inside it?  Cause the `FrameTransport` object itself does everything under a lock.